### PR TITLE
Added checks on distributed vectors for interpolate

### PR DESCRIPTION
**Pull request by GivAlz**
Timo and I found a little nasty bug (actually the issue of yesterday was real...only it was difficult to find not in test mode).

@tjhei as soon as you comment I shall fix and this one should be done :) Thank you!!!

### CODE REVIEW

**Comment by tjhei** [comment]
https://github.com/dealii/dealii/blob/master/include/deal.II/numerics/vector_tools.templates.h#L8914

**Comment by tjhei** [comment]
```
template <class vec>
void check() {}

template <class vec> enable_if <is_base_of<BlockVectorBase<T, vec>>
void check() {
for each block i: check(vec.block(i));
}

template <class vec> enable_if <is_serial_vector<vec>==false>
void check() {
check has_ghost_elements()....
}



```

**Comment by tjhei** [comment]
grab me tomorrow and we will figure it out together.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 1]
```diff
@@ -265,6 +265,41 @@ namespace VectorTools
...
+        // If the vector is distributed, the following checks
+        // try to prevent problems on ghost cells
+        Assert(!vt.has_ghost_elements(),
+               ExcMessage("Vector has no write access on ghost cells"));
+
+        // Cheking for overlapping values on ghost cells:
+        // this test is not 100% accurate, but should catch most of the
+        // problems
+        const auto n_global_elements =
+          Utilities::MPI::sum(vt.local_size() + vt.n_ghost_entries(),
```
Comment (masterleinad) [inline include/deal.II/numerics/vector_tools.templates.h]: `n_ghost_entries`  only exists for `LinearAlgebra::distributed::Vector`.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 2]
```diff
@@ -265,6 +265,41 @@ namespace VectorTools
...
 
 
+    template <typename VectorType>
+    typename std::enable_if<dealii::is_serial_vector<VectorType>::value ==
+                          true>::type
+            check_ghost(VectorType &vt) {}
+
+    template <typename VectorType>
+    typename std::enable_if<dealii::is_serial_vector<VectorType>::value == false
+            and std::is_base_of<BlockVectorBase<typename VectorType>,VectorType>::value == false >::type
```
Comment (masterleinad) [inline include/deal.II/numerics/vector_tools.templates.h]: You can use `IsBlockVector<VectorType>::value` instead.

File: include/deal.II/numerics/vector_tools.templates.h (hunk 1 above)
Comment (GivAlz) [reply include/deal.II/numerics/vector_tools.templates.h]: Das ist ein großes, haariges graues Problem... just joking, thank you for pointing this out! We shall think of something...

**Comment by masterleinad** [comment]
What is the problem you are trying to fix here anyway?

**Comment by GivAlz** [comment]
Currently the user can interpolate using distributed vectors in an incorrect way (see the test). The test in this PR (launched as test) fails because in truth a 0/0 operation is happening when interpolating on the some entries (ghost/artificial entries, I guess, Timo know better), but it will launch if you compile with a separate cmake file...and, oftentimes, users will get lucky and everything will work. Thus this should warn the user and throw a readable error message (instead of a weird floating point exception).

( To do it properly the user has first to activate writing on ghost cells, interpolate, exchange data on ghost entries and then deactivate writing (which activates reading). )

**Comment by masterleinad** [comment]
Is this only an issue with `LinearAlgebra::distributed::Vector`?

**Comment by masterleinad** [comment]
I would instead only write to the entry if `std::abs(weight)>0.`.

File: tests/vector_tools/interpolate_distributed_01.cc [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  parallel::distributed::Triangulation<dim> cube_d(mpi_communicator);
+  GridGenerator::hyper_cube(cube_d);
+  cube_d.refine_global(ref_cube);
+  deallog << "Initializing the dof handler" << std::endl;
+  DoFHandler<dim> dof_handler_d(cube_d);
+  dof_handler_d.distribute_dofs(fe);
+
+  deallog << "Interpolating data" << std::endl;
+  LinearAlgebra::distributed::Vector<double> data_d(dof_handler_d.n_dofs());
```
Comment (kronbichler) [inline tests/vector_tools/interpolate_distributed_01.cc]: I have not looked into this more closely yet, but shouldn't you provide a partitioning at this point? If you want to use the vector in parallel, you would specify the locally owned range of the DoFHandler and the relevant dofs as ghost (for example). The tests in themselves seem to make sense, but I do not understand the test (yet).

**Comment by GivAlz** [comment]
Ok, so now I use the suggestion of @masterleinad  and check for the weights to be positive. @kronbichler  I stumbled in a situation where the operation 0/0 is done, and the tester was complaining about this.

Btw, I am sorry, I think I made a bit of a mess with a previous commit...I'll try to fix this. But, other than that, this should be ok...

File: doc/news/changes/minor/20180728GiovanniAlzettaTimoHeister [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Corrected potential bug: VectorTools::interpolate now checks for weights to be positive.
+<br>
+(Giovanni Alzetta, Timo Heister, 2018/07/24)
```
Comment (masterleinad) [inline doc/news/changes/minor/20180728GiovanniAlzettaTimoHeister]: At least, this is not accurate anymore. Now, we just set the values for all the components we know about.

**Pull request closed without merging**